File: scripts/algorithms/M/Maximum Genetic Difference Query/Maximum Genetic Difference Query.cpp
/* 
    Time: O(q+n)
    Space: O(q+n)
    Tag: Trie, N-ary Tree
    Difficulty: H
*/
struct Node {
    int count;
    vector<Node *> children;
    Node() {
        children.resize(2, NULL);
        count = 0;
    }
};

class Trie {
    Node *root;

public:
    Trie() {
        root = new Node();
    }

    void insert(int num) {
        Node *cur = root;
        for (int i = 18; i >= 0; i--) {
            bool bit = (num & (1 << i));
            if (!cur->children[bit]) cur->children[bit] = new Node();
            cur = cur->children[bit];
            cur->count++;
        }
    }

    void remove(int num) {
        Node *cur = root;
        for (int i = 18; i >= 0; i--) {
            bool bit = (num & (1 << i));
            cur = cur->children[bit];
            cur->count--;
        }
    }

    int search(int num) {
        int res = 0;
        Node *cur = root;
        for (int i = 18; i >= 0; i--) {
            bool bit = (num & (1 << i));
            if (cur->children[!bit] && cur->children[!bit]->count > 0) {
                res += (1 << i);
                cur = cur->children[!bit];
            } else {
                cur = cur->children[bit];
            }
        }
        return res;
    }
};

class Solution {

    vector<vector<pair<int, int>>> mp;
    Trie *t;
    vector<int> res;

    void dfs(vector<int> gr[], int cur) {
        t->insert(cur);
        for (auto &child : mp[cur]) {
            res[child.second] = t->search(child.first);
        }
        for (int &child : gr[cur]) {
            dfs(gr, child);
        }
        t->remove(cur);
    }

public:
    vector<int> maxGeneticDifference(vector<int> &parents, vector<vector<int>> &queries) {
        t = new Trie();

        int n = parents.size();
        vector<int> gr[n];
        res.resize(queries.size());

        int root = -1;

        for (int i = 0; i < n; i++) {
            if (parents[i] == -1)
                root = i;
            else
                gr[parents[i]].push_back(i);
        }
        mp.resize(n);
        for (int i = 0; i < queries.size(); i++) {
            mp[queries[i][0]].push_back({queries[i][1], i});
        }

        dfs(gr, root);

        return res;
    }
};

File: scripts/algorithms/M/Maximum Genetic Difference Query/Maximum Genetic Difference Query.py
# Runtime: 6663 ms (Top 46.4%) | Memory: 339.09 MB (Top 7.1%)

class Solution:
    def maxGeneticDifference(self, parents: List[int], queries: List[List[int]]) -> List[int]:
        adj_map = collections.defaultdict(set)
        root = None
        for i, node in enumerate(parents):
            if node == -1:
                root = i
            else:
                adj_map[node].add(i)
        queries_map = collections.defaultdict(set)

        for q in queries:
            queries_map[q[0]].add(q[1])
        self.res_map = {}
        def helperDFS(curr_root,prefix_map):

            if curr_root in queries_map:
                for val in queries_map[curr_root]:
                    bin_rep = format(val, '020b')
                    best_bin = ""
                    print(bin_rep)
                    for i in range(20):
                        if prefix_map[best_bin+str(1-int(bin_rep[i]))]:
                            best_bin += str(1-int(bin_rep[i]))
                        else:
                            best_bin += bin_rep[i]
                    self.res_map[(curr_root,val)] = int(best_bin,2) ^ val 
            for child in adj_map[curr_root]:
                bin_rep = format(child, '020b')
                for i in range(1, len(bin_rep)+1):
                    prefix_map[bin_rep[0:i]].add(child)
                helperDFS(child, prefix_map)
                for i in range(1, len(bin_rep)+1):
                    prefix_map[bin_rep[0:i]].remove(child)

        initial_prefixmap = collections.defaultdict(set)
        root_rep = format(root, '020b')
        for i in range(1,21):
            initial_prefixmap[root_rep[0:i]].add(root)
        helperDFS(root, initial_prefixmap)
        res = []
        for q in queries:
            res.append(self.res_map[(q[0],q[1])])
        return res
            
